Ab/initial changes for mspc

abarciauskas-bgse opened the pull request:
This PR primarily:
* removes WMTS code + documentation
* Adds planetary computer library and code to sign STAC asset urls

I made this a draft primarily because I'm still thinking what could be the best approach for conditionally signing requests with the planetary computer library.

vincentsarago commented on docs/src/endpoints/items_endpoints.md:
```diff
@@ -13,7 +13,6 @@ The `Item` endpoints are created using TiTiler's [MultiBaseTilerFactory](https:/
 | `POST` | `/collections/{collection_id}/items/{item_id}/statistics`                                                   | GeoJSON ([Statistics][multistats_geojson_model]) | return assets statistics for a GeoJSON (merged)
 | `GET`  | `/collections/{collection_id}/items/{item_id}/tiles[/{TileMatrixSetId}]/{z}/{x}/{y}[@{scale}x][.{format}]`  | image/bin                                        | create a web map tile image from assets
 | `GET`  | `/collections/{collection_id}/items/{item_id}[/{TileMatrixSetId}]/tilejson.json`                            | JSON ([TileJSON][tilejson_model])                | return a Mapbox TileJSON document
-| `GET`  | `/collections/{collection_id}/items/{item_id}[/{TileMatrixSetId}]/WMTSCapabilities.xml`                     | XML                                              | return OGC WMTS Get Capabilities
```
the `items` endpoints comes from titiler `MultiBaseTilerFactory` so it will still be present

abarciauskas-bgse replied:
change was reverted

vincentsarago commented on titiler/stacapi/settings.py:
```diff
@@ -77,6 +77,7 @@ class STACAPISettings(BaseSettings):
     """STAC API settings"""
 
     stac_api_url: str
+    mspc_default_api_url: str = "https://planetarycomputer.microsoft.com/api/stac/v1"
```
I think it's fine having `stac_api_url` to default to `https://planetarycomputer.microsoft.com/api/stac/v1`

vincentsarago commented on titiler/stacapi/reader.py:
```diff
@@ -82,6 +84,15 @@ def _get_asset_info(self, asset: str) -> AssetInfo:
+        # 2. Set a boolean in settings, something like "SIGN_REQUESTS"
+        # 3. Just assume all requests are to MS PC API and sign them. Stub out this function in tests and/or when ENV=test
+        if stac_api_config.stac_api_url == stac_api_config.mspc_default_api_url:
+            url = planetary_computer.sign(url)
```
FYI this is how it's done in MC planetary computer https://github.com/microsoft/planetary-computer-apis/blob/main/pctiler/pctiler/reader.py#L43-L59

vincentsarago replied:
to avoid having a new settings, we could simply check if the asset url is on MS datastore

abarciauskas-bgse replied:
>FYI this is how it's done in MC planetary computer https://github.com/microsoft/planetary-computer-apis/blob/main/pctiler/pctiler/reader.py#L43-L59

Nice, thanks for sharing that. It looks like we would not use this code exactly because it relies on a `CollectionConfigTable`

>to avoid having a new settings, we could simply check if the asset url is on MS datastore

👍🏽 it does appear we can re-use some of https://github.com/microsoft/planetary-computer-apis/blob/main/pccommon/pccommon/cdn.py#L8 to do this. I noticed that this code determines if there is an azureedge configured for the data, but I don't think there is anyway for this application to know about that? Since that code relies on (I think) some redis cache? https://github.com/microsoft/planetary-computer-apis/blob/main/pccommon/pccommon/config/core.py#L28

hrodmn reviewed (commented):
@abarciauskas-bgse I'm sorry I missed your request for a review - didn't have my notifications set up yet! That is fixed now.

We discussed this in our call today but one other way to sign the assets is to use the `modifier` parameter in the `pystac.ItemSearch` (or `pystac.Client`) class where we use it (backend.py, dependencies.py, factory.py I think)

```python
ItemSearch(..., modifier=planetary_computer.sign_inplace)
```

I don't know if it would be more efficient than checking each asset url, but maybe it would be!

abarciauskas-bgse commented:
@hrodmn I used the ItemSearch modifier option signing method. I verified it didn't break when non-MS PC results are returned from search. So this does seem cleaner than including a regex to check URLs against.

@vincentsarago are there any other initial changes you think we should make? For example, I am not sure if the models in models.py are being used apart from the `Link` model and I think they were used as part of the WMTS endpoints. Do you think it's safe to remove all of the models apart from `Link`?

vincentsarago commented on titiler/stacapi/backend.py:
```diff
@@ -258,6 +259,7 @@ def get_assets(
             f"{self.url}/search",
             stac_io=stac_api_io,
             **params,
+            modifier=pc.sign_inplace,
```
❤️

vincentsarago commented on titiler/stacapi/settings.py:
```diff
@@ -67,7 +68,7 @@ class RetrySettings(BaseSettings):
 
     model_config = {
-        "env_prefix": "TITILER_STACAPI_API_",
+        "env_prefix": "TITILER_STACAPI_RETRY_",
```
why do we need this change?

abarciauskas-bgse replied:
some of these `env_prefix`s were the same, which seems like it could cause a bug if one of these classes both had the same attribute name so I thought we could rename the env_prefixes to match the class name. What do you think?

vincentsarago replied:
It's not necessary, pydantic basettings is smart enough to handle this. 

they are distributed into multiple classes because they are not used at the same places

abarciauskas-bgse replied:
ok reverted those changes

vincentsarago commented on titiler/stacapi/stac_reader.py:
```diff
@@ -12,9 +14,10 @@
+from titiler.stacapi.settings import STACAPISettings, STACSettings
 
 stac_config = STACSettings()
+stac_api_config = STACAPISettings()
```
This setting is not used?

abarciauskas-bgse replied:
thanks for noticing this, I had misconfigured the pre-commit linter so missed it!

vincentsarago commented:
@abarciauskas-bgse yes, you can remove https://github.com/developmentseed/titiler-stacapi-mspc/blob/ab/initial-changes-for-mspc/titiler/stacapi/models.py#L88-L115

abarciauskas-bgse commented:
@vincentsarago I moved this from draft to ready for review as I have addressed the outstanding questions I had. Let me know if you think any other changes should be made for this initial PR.
